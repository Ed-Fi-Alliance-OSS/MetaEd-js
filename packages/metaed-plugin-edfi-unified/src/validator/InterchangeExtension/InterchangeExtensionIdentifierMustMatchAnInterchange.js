// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.interchangeExtension.forEach(interchangeExtension => {
    if (metaEd.entity.interchange.has(interchangeExtension.metaEdName)) return;
    failures.push({
      validatorName: 'InterchangeExtensionIdentifierMustMatchAnInterchange',
      category: 'error',
      message: `Interchange additions ${interchangeExtension.metaEdName} does not match any declared Interchange.`,
      sourceMap: interchangeExtension.sourceMap.metaEdName,
      fileMap: null,
    });
  });

  return failures;
}
