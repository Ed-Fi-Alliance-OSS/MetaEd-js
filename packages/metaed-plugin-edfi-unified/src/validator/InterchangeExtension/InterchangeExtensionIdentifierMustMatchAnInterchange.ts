// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { MetaEdEnvironment, ValidationFailure, Namespace } from '@edfi/metaed-core';
import { getEntityFromNamespaceChain } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.interchangeExtension.forEach((interchangeExtension) => {
      if (
        getEntityFromNamespaceChain(
          interchangeExtension.baseEntityName,
          interchangeExtension.baseEntityNamespaceName,
          namespace,
          'interchange',
        ) == null
      ) {
        failures.push({
          validatorName: 'InterchangeExtensionIdentifierMustMatchAnInterchange',
          category: 'error',
          message: `Interchange additions ${interchangeExtension.metaEdName} does not match any declared Interchange in namespace ${interchangeExtension.baseEntityNamespaceName}.`,
          sourceMap: interchangeExtension.sourceMap.metaEdName,
          fileMap: null,
        });
      }
    });
  });
  return failures;
}
