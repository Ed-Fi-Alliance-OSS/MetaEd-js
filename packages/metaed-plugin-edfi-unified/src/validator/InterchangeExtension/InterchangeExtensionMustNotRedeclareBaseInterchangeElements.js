// @flow
import type { MetaEdEnvironment, ValidationFailure, ModelBase } from 'metaed-core';
import { getEntityForNamespaces } from 'metaed-core';
import { failInterchangeExtensionPropertyRedeclarations } from '../ValidatorShared/FailInterchangeExtensionPropertyRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.interchangeExtension.forEach(interchangeExtension => {
      const extendedEntity: ?ModelBase = getEntityForNamespaces(
        interchangeExtension.metaEdName,
        [...namespace.dependencies],
        'interchange',
      );
      if (extendedEntity == null) return;

      failInterchangeExtensionPropertyRedeclarations(
        'InterchangeExtensionMustNotRedeclareBaseInterchangeElements',
        'elements',
        interchangeExtension,
        extendedEntity,
        failures,
      );
    });
  });
  return failures;
}
