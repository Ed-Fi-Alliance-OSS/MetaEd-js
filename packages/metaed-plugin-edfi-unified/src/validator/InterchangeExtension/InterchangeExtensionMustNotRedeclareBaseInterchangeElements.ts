import { MetaEdEnvironment, ValidationFailure, Interchange, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';
import { failInterchangeExtensionPropertyRedeclarations } from '../ValidatorShared/FailInterchangeExtensionPropertyRedeclarations';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.interchangeExtension.forEach(interchangeExtension => {
      const extendedEntity: Interchange | null = getEntityFromNamespaceChain(
        interchangeExtension.baseEntityName,
        interchangeExtension.baseEntityNamespaceName,
        namespace,
        'interchange',
      ) as Interchange | null;

      if (extendedEntity == null) return;

      failInterchangeExtensionPropertyRedeclarations(
        'InterchangeExtensionMustNotRedeclareBaseInterchangeElements',
        'elements',
        interchangeExtension,
        extendedEntity,
        failures,
      );
    });
  });
  return failures;
}
