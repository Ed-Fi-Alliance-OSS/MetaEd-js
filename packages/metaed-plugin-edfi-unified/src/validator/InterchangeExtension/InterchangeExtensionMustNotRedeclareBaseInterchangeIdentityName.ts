import { MetaEdEnvironment, ValidationFailure, Interchange, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';
import { failInterchangeExtensionPropertyRedeclarations } from '../ValidatorShared/FailInterchangeExtensionPropertyRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.interchangeExtension.forEach(interchangeExtension => {
      const extendedEntity: Interchange | null = getEntityFromNamespaceChain(
        interchangeExtension.baseEntityName,
        interchangeExtension.baseEntityNamespaceName,
        namespace,
        'interchange',
      ) as Interchange | null;

      if (extendedEntity == null) return;
      failInterchangeExtensionPropertyRedeclarations(
        'InterchangeExtensionMustNotRedeclareBaseInterchangeIdentityName',
        'identityTemplates',
        interchangeExtension,
        extendedEntity,
        failures,
      );
    });
  });
  return failures;
}
