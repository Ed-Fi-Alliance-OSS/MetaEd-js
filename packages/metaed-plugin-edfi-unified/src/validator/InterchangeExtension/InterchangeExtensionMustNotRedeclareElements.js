// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../metaed-core/index';
import { failInterchangeItemRedeclarations } from '../ValidatorShared/FailInterchangeItemRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.interchangeExtension.forEach(interchangeExtension => {
    if (interchangeExtension.elements.length === 0) return;
    failInterchangeItemRedeclarations(
      'InterchangeExtensionMustNotRedeclareElements', 'element', interchangeExtension, interchangeExtension.elements, failures);
  });
  return failures;
}
