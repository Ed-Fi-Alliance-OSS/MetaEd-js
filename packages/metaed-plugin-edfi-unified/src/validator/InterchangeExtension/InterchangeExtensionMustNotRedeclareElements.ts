import { MetaEdEnvironment, ValidationFailure, Namespace } from '@edfi/metaed-core';
import { failInterchangeItemRedeclarations } from '../ValidatorShared/FailInterchangeItemRedeclarations';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.interchangeExtension.forEach((interchangeExtension) => {
      if (interchangeExtension.elements.length === 0) return;
      failInterchangeItemRedeclarations(
        'InterchangeExtensionMustNotRedeclareElements',
        'element',
        interchangeExtension,
        interchangeExtension.elements,
        failures,
      );
    });
  });
  return failures;
}
