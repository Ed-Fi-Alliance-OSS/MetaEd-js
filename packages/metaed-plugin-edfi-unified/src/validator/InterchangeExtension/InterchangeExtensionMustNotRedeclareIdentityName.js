// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { failInterchangeItemRedeclarations } from '../ValidatorShared/FailInterchangeItemRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.interchangeExtension.forEach(interchangeExtension => {
    if (interchangeExtension.identityTemplates.length === 0) return;
    failInterchangeItemRedeclarations(
      'InterchangeExtensionMustNotRedeclareIdentityName', 'identity template', interchangeExtension, interchangeExtension.identityTemplates, failures);
  });
  return failures;
}
