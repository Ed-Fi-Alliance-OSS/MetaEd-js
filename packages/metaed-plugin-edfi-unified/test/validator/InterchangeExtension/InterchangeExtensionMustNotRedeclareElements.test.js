// @flow
import { newMetaEdEnvironment, MetaEdTextBuilder, InterchangeBuilder } from 'metaed-core';
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { validate } from '../../../src/validator/InterchangeExtension/InterchangeExtensionMustNotRedeclareElements';

describe('when validating interchange extension element has different names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartInterchangeExtension('InterchangeName')
      .withDomainEntityElement('DomainEntityElementName1')
      .withDomainEntityElement('DomainEntityElementName2')
      .withEndInterchangeExtension()
      .withEndNamespace()
      .sendToListener(new InterchangeBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should build one interchange extension', () => {
    expect(metaEd.entity.interchangeExtension.size).toBe(1);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating interchange element has duplicate names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityElementName: string = 'DomainEntityElementName';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartInterchangeExtension('InterchangeName')
      .withDomainEntityElement(domainEntityElementName)
      .withDomainEntityElement(domainEntityElementName)
      .withEndInterchangeExtension()
      .withEndNamespace()
      .sendToListener(new InterchangeBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should build one interchange extension', () => {
    expect(metaEd.entity.interchangeExtension.size).toBe(1);
  });

  it('should have validation failures', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('InterchangeExtensionMustNotRedeclareElements');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when validating interchange element has duplicate names should have no validation failures -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when validating interchange element has duplicate names should have no validation failures -> sourceMap');
  });
});

describe('when validating interchange element has multiple duplicate names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityElementName1: string = 'DomainEntityElementName1';
  const domainEntityElementName2: string = 'DomainEntityElementName2';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartInterchangeExtension('InterchangeName')
      .withDomainEntityElement(domainEntityElementName1)
      .withDomainEntityElement(domainEntityElementName1)
      .withDomainEntityElement(domainEntityElementName2)
      .withDomainEntityElement(domainEntityElementName2)
      .withEndInterchangeExtension()
      .withEndNamespace()
      .sendToListener(new InterchangeBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should build one interchange extension', () => {
    expect(metaEd.entity.interchangeExtension.size).toBe(1);
  });

  it('should have validation failures', () => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('InterchangeExtensionMustNotRedeclareElements');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when validating interchange element has multiple duplicate names should have validation failures -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when validating interchange element has multiple duplicate names should have validation failures -> sourceMap');
    expect(failures[1].validatorName).toBe('InterchangeExtensionMustNotRedeclareElements');
    expect(failures[1].category).toBe('error');
    expect(failures[1].message).toMatchSnapshot('when validating interchange element has multiple duplicate names should have validation failures -> message');
    expect(failures[1].sourceMap).toMatchSnapshot('when validating interchange element has multiple duplicate names should have validation failures -> sourceMap');
  });
});
