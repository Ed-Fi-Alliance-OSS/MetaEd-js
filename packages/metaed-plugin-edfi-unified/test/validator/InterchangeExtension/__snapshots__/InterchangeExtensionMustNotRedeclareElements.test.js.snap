// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating interchange element has duplicate names should have validation failures: when validating interchange element has duplicate names should have no validation failures -> message 1`] = `"Interchange InterchangeName redeclares element DomainEntityElementName."`;

exports[`when validating interchange element has duplicate names should have validation failures: when validating interchange element has duplicate names should have no validation failures -> sourceMap 1`] = `
Object {
  "column": 4,
  "line": 3,
  "tokenText": "domain entity",
}
`;

exports[`when validating interchange element has multiple duplicate names should have validation failures: when validating interchange element has multiple duplicate names should have validation failures -> message 1`] = `"Interchange InterchangeName redeclares element DomainEntityElementName1."`;

exports[`when validating interchange element has multiple duplicate names should have validation failures: when validating interchange element has multiple duplicate names should have validation failures -> message 2`] = `"Interchange InterchangeName redeclares element DomainEntityElementName2."`;

exports[`when validating interchange element has multiple duplicate names should have validation failures: when validating interchange element has multiple duplicate names should have validation failures -> sourceMap 1`] = `
Object {
  "column": 4,
  "line": 3,
  "tokenText": "domain entity",
}
`;

exports[`when validating interchange element has multiple duplicate names should have validation failures: when validating interchange element has multiple duplicate names should have validation failures -> sourceMap 2`] = `
Object {
  "column": 4,
  "line": 5,
  "tokenText": "domain entity",
}
`;
